File: smartdesk-frontend/src/services/ticketService.ts
import { apiClient } from './apiClient';
import { Ticket, CreateTicketRequest, UpdateTicketRequest } from '../types';

export const ticketService = {
    // Get all tickets - SUPER_ADMIN gets all, CUSTOMER gets own tickets, others get company tickets
    async getTickets(params?: {
        page?: number;
        size?: number;
        status?: string;
        priority?: string;
        search?: string;
        companyId?: number;
        userRole?: string;
        userId?: number;
    }) {
        // Use search endpoint with companyId
        if (params?.search || params?.status || params?.priority) {
            const searchParams: any = {
                companyId: params.companyId || 1,
                page: params.page || 0,
                size: params.size || 20
            };
            if (params.search) searchParams.q = params.search;
            if (params.status) searchParams.status = params.status;
            if (params.priority) searchParams.priority = params.priority;
            
            const response = await apiClient.get('/tickets/search', { params: searchParams });
            console.log('🎫 TicketService search response:', response);
            return {
                data: response.data?.tickets || [],
                totalElements: response.data?.totalElements || 0,
                totalPages: response.data?.totalPages || 0
            };
        } else {
            // SUPER_ADMIN gets all tickets
            if (params?.userRole === 'SUPER_ADMIN') {
                console.log('🎫 SUPER_ADMIN detected - fetching ALL tickets');
                const response = await apiClient.get('/tickets/all', { 
                    params: { 
                        page: params?.page || 0, 
                        size: params?.size || 20 
                    } 
                });
                console.log('🎫 TicketService ALL tickets response:', response);
                return {
                    data: response.data?.tickets || [],
                    totalElements: response.data?.totalElements || 0,
                    totalPages: response.data?.totalPages || 0
                };
            } 
            // CUSTOMER gets only their own tickets
            else if (params?.userRole === 'CUSTOMER') {
                console.log('🎫 CUSTOMER detected - fetching my tickets');
                const response = await apiClient.get('/tickets/my', { 
                    params: { 
                        page: params?.page || 0, 
                        size: params?.size || 20,
                        sort: 'createdAt,desc'  // Yeni ticket'lar en üstte
                    } 
                });
                console.log('🎫 TicketService my tickets response:', response);
                return {
                    data: response.data?.tickets || [],
                    totalElements: response.data?.totalElements || 0,
                    totalPages: response.data?.totalPages || 0
                };
            }
            // Others (MANAGER, AGENT) get company tickets
            else {
                console.log('🎫 MANAGER/AGENT detected - fetching company tickets for companyId:', params?.companyId);
                const response = await apiClient.get(`/tickets/company/${params?.companyId || 1}`, { 
                    params: { 
                        page: params?.page || 0, 
                        size: params?.size || 20,
                        sort: 'createdAt,desc'  // Yeni ticket'lar en üstte
                    } 
                });
                console.log('🎫 TicketService company tickets response:', response);
                // Backend returns custom response with "tickets" array (not "content")
                return {
                    data: response.data?.tickets || [],
                    totalElements: response.data?.totalElements || 0,
                    totalPages: response.data?.totalPages || 0
                };
            }
        }
    },

    // Get ticket by ID
    async getTicketById(id: number) {
        const response = await apiClient.get(`/tickets/${id}`);
        return response.data;
    },

    // Create new ticket
    async createTicket(ticketData: CreateTicketRequest) {
        const response = await apiClient.post('/tickets', ticketData);
        return response.data;
    },

    // Update ticket
    async updateTicket(id: number, ticketData: UpdateTicketRequest) {
        const response = await apiClient.put(`/tickets/${id}`, ticketData);
        return response.data;
    },

    // Change ticket status
    async changeTicketStatus(id: number, status: string) {
        const response = await apiClient.patch(`/tickets/${id}/status`, { status });
        return response.data;
    },

    // Update ticket status (alias for compatibility)
    async updateTicketStatus(id: number, status: string) {
        return this.changeTicketStatus(id, status);
    },

    // Change ticket priority
    async changeTicketPriority(id: number, priority: string) {
        const response = await apiClient.patch(`/tickets/${id}/priority`, { priority });
        return response.data;
    },

    // Assign ticket to agent
    async assignTicket(id: number, agentId: number) {
        const response = await apiClient.patch(`/tickets/${id}/assign`, { agentId });
        return response.data;
    },

    // Close ticket
    async closeTicket(id: number, resolution?: string) {
        const response = await apiClient.patch(`/tickets/${id}/close`, { resolution });
        return response.data;
    },

    // Get ticket comments
    async getTicketComments(id: number, includeInternal = false) {
        const response = await apiClient.get(`/tickets/${id}/comments`, {
            params: { includeInternal }
        });
        return response.data;
    },

    // Add comment to ticket
    async addComment(id: number, commentData: {
        content: string;
        isInternal?: boolean;
    }) {
        const response = await apiClient.post(`/tickets/${id}/comments`, commentData);
        return response.data;
    },

    // Add attachment to ticket
    async addAttachment(id: number, file: File) {
        const formData = new FormData();
        formData.append('file', file);

        const response = await apiClient.post(`/tickets/${id}/attachments`, formData, {
            headers: {
                'Content-Type': 'multipart/form-data',
            },
        });
        return response.data;
    },

    // Get tickets by customer
    async getTicketsByCustomer(customerId: number, params?: {
        page?: number;
        size?: number;
    }) {
        const response = await apiClient.get(`/tickets/customer/${customerId}`, { params });
        return response.data;
    },

    // Get tickets by agent
    async getTicketsByAgent(agentId: number, params?: {
        page?: number;
        size?: number;
    }) {
        const response = await apiClient.get(`/tickets/agent/${agentId}`, { params });
        return response.data;
    },

    // Search tickets
    async searchTickets(params: {
        companyId: number;
        q?: string;
        status?: string;
        priority?: string;
        departmentId?: number;
        agentId?: number;
        customerId?: number;
        page?: number;
        size?: number;
    }) {
        const response = await apiClient.get('/tickets/search', { params });
        return response.data;
    },

    // Get ticket statistics
    async getTicketStats(companyId: number) {
        const response = await apiClient.get('/tickets/stats', {
            params: { companyId }
        });
        return response.data;
    },

    // Get active tickets
    async getActiveTickets(companyId: number, params?: {
        page?: number;
        size?: number;
    }) {
        const response = await apiClient.get('/tickets/active', {
            params: { companyId, ...params }
        });
        return response.data;
    },

    // Get unassigned tickets
    async getUnassignedTickets(companyId: number) {
        const response = await apiClient.get('/tickets/unassigned', {
            params: { companyId }
        });
        return response.data;
    },

    // Get tickets at SLA risk
    async getTicketsAtSlaRisk(companyId: number) {
        const response = await apiClient.get('/tickets/sla-risk', {
            params: { companyId }
        });
        return response.data;
    },

    // Add satisfaction rating
    async addSatisfactionRating(id: number, rating: number, feedback?: string) {
        const response = await apiClient.post(`/tickets/${id}/satisfaction`, {
            rating,
            feedback
        });
        return response.data;
    },

    // Get ticket creation trends
    async getTicketCreationTrend(companyId: number, days = 30) {
        const response = await apiClient.get('/tickets/trends/creation', {
            params: { companyId, days }
        });
        return response.data;
    },

    // ============ APPROVAL WORKFLOW METHODS ============

    // Resolve ticket for manager approval
    async resolveTicket(id: number, resolutionSummary?: string) {
        const response = await apiClient.patch(`/tickets/${id}/resolve`, {
            resolutionSummary: resolutionSummary || ''
        });
        return response.data;
    },

    // Manager approval
    async approveByManager(id: number, approvalComment?: string) {
        const response = await apiClient.patch(`/tickets/${id}/approve-manager`, {
            approvalComment: approvalComment || ''
        });
        return response.data;
    },

    // Admin final approval
    async approveByAdmin(id: number, finalComment?: string) {
        const response = await apiClient.patch(`/tickets/${id}/approve-admin`, {
            finalComment: finalComment || ''
        });
        return response.data;
    },

    // Reject approval
    async rejectApproval(id: number, rejectionReason: string) {
        const response = await apiClient.patch(`/tickets/${id}/reject-approval`, {
            rejectionReason
        });
        return response.data;
    }
}; 

File: src/main/java/com/example/smartdeskbackend/service/TicketService.java
package com.example.smartdeskbackend.service;

import com.example.smartdeskbackend.dto.request.ticket.CreateTicketRequest;
import com.example.smartdeskbackend.dto.request.ticket.UpdateTicketRequest;
import com.example.smartdeskbackend.dto.request.ticket.AssignTicketRequest;
import com.example.smartdeskbackend.dto.request.ticket.AddCommentRequest;
import com.example.smartdeskbackend.dto.response.ticket.TicketResponse;
import com.example.smartdeskbackend.dto.response.ticket.TicketDetailResponse;
import com.example.smartdeskbackend.dto.response.ticket.TicketCommentResponse;
import com.example.smartdeskbackend.enums.TicketStatus;
import com.example.smartdeskbackend.enums.TicketPriority;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;
import org.springframework.web.multipart.MultipartFile;

import java.time.LocalDateTime;
import java.util.List;
import java.util.Map;

/**
 * Ticket service interface
 * Ticket yönetimi için tüm business logic operasyonlarını tanımlar
 */
public interface TicketService {

    // ============ TEMEL CRUD OPERASYONLARI ============

    /**
     * Ticket detaylarını ID ile getir
     * @param id Ticket ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse getTicketById(Long id);

    /**
     * Ticket numarası ile ticket getir
     * @param ticketNumber Unique ticket number (TK-YYYYMMDD-XXXX format)
     * @return TicketDetailResponse
     */
    TicketDetailResponse getTicketByNumber(String ticketNumber);

    /**
     * Şirketteki tüm ticketları sayfalı olarak getir
     * @param companyId Şirket ID
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> getTicketsByCompany(Long companyId, Pageable pageable);

    /**
     * Tüm ticketları getir (SUPER_ADMIN için)
     * @param pageable Pagination bilgisi
     * @return Tüm ticketlar
     */
    Page<TicketResponse> getAllTickets(Pageable pageable);

    /**
     * Belirli müşterinin ticketlarını getir
     * @param customerId Müşteri ID
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> getTicketsByCustomer(Long customerId, Pageable pageable);

    /**
     * Belirli agent'a atanmış ticketları getir
     * @param agentId Agent ID
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> getTicketsByAgent(Long agentId, Pageable pageable);

    /**
     * CUSTOMER kullanıcının kendi ticketlarını getir (User ID ile)
     * @param userId User ID (CUSTOMER role)
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> getTicketsByUserId(Long userId, Pageable pageable);

    // ============ ARAMA ve FİLTRELEME ============

    /**
     * Ticket arama (title, description, ticket number)
     * @param companyId Şirket ID
     * @param searchTerm Arama terimi
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> searchTickets(Long companyId, String searchTerm, Pageable pageable);

    /**
     * Gelişmiş filtreleme ile ticket getirme
     * @param companyId Şirket ID
     * @param status Ticket durumu (opsiyonel)
     * @param priority Ticket önceliği (opsiyonel)
     * @param departmentId Departman ID (opsiyonel)
     * @param agentId Agent ID (opsiyonel)
     * @param customerId Müşteri ID (opsiyonel)
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> getTicketsWithFilters(Long companyId, TicketStatus status,
                                               TicketPriority priority, Long departmentId,
                                               Long agentId, Long customerId, Pageable pageable);

    /**
     * Aktif ticketları getir (NEW, OPEN, IN_PROGRESS, PENDING)
     * @param companyId Şirket ID
     * @param pageable Sayfalama bilgisi
     * @return Page<TicketResponse>
     */
    Page<TicketResponse> getActiveTickets(Long companyId, Pageable pageable);

    /**
     * Atanmamış ticketları getir
     * @param companyId Şirket ID
     * @return List<TicketResponse>
     */
    List<TicketResponse> getUnassignedTickets(Long companyId);

    /**
     * SLA riski olan ticketları getir
     * @param companyId Şirket ID
     * @return List<TicketResponse>
     */
    List<TicketResponse> getTicketsAtSlaRisk(Long companyId);

    // ============ TİCKET OLUŞTURMA ve GÜNCELLEME ============

    /**
     * Yeni ticket oluştur
     * @param request Ticket oluşturma isteği
     * @return TicketDetailResponse
     */
    TicketDetailResponse createTicket(CreateTicketRequest request);

    /**
     * Ticket bilgilerini güncelle
     * @param id Ticket ID
     * @param request Güncelleme isteği
     * @return TicketDetailResponse
     */
    TicketDetailResponse updateTicket(Long id, UpdateTicketRequest request);

    // ============ TİCKET DURUM YÖNETİMİ ============

    /**
     * Ticket durumunu değiştir
     * @param id Ticket ID
     * @param newStatus Yeni durum
     * @param userId İşlemi yapan kullanıcı ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse changeTicketStatus(Long id, TicketStatus newStatus, Long userId);

    /**
     * Ticket'ı agent'a ata
     * @param id Ticket ID
     * @param request Atama isteği
     * @return TicketDetailResponse
     */
    TicketDetailResponse assignTicket(Long id, AssignTicketRequest request);

    /**
     * Ticket'ı otomatik olarak uygun agent'a ata
     * @param id Ticket ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse autoAssignTicket(Long id);

    /**
     * Ticket önceliğini değiştir
     * @param id Ticket ID
     * @param newPriority Yeni öncelik
     * @param userId İşlemi yapan kullanıcı ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse changeTicketPriority(Long id, TicketPriority newPriority, Long userId);

    /**
     * Ticket'ı escalate et (öncelik artırma)
     * @param id Ticket ID
     * @param userId İşlemi yapan kullanıcı ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse escalateTicket(Long id, Long userId);

    /**
     * Ticket'ı kapat
     * @param id Ticket ID
     * @param resolutionSummary Çözüm özeti
     * @param userId İşlemi yapan kullanıcı ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse closeTicket(Long id, String resolutionSummary, Long userId);

    // ============ YORUM ve EK DOSYA YÖNETİMİ ============

    /**
     * Ticket'a yorum ekle
     * @param ticketId Ticket ID
     * @param request Yorum ekleme isteği
     * @return TicketCommentResponse
     */
    TicketCommentResponse addComment(Long ticketId, AddCommentRequest request);

    /**
     * Ticket yorumlarını getir
     * @param ticketId Ticket ID
     * @param includeInternal Internal yorumları da dahil et
     * @return List<TicketCommentResponse>
     */
    List<TicketCommentResponse> getTicketComments(Long ticketId, boolean includeInternal);

    /**
     * Ticket'a dosya eki ekle
     * @param ticketId Ticket ID
     * @param file Yüklenecek dosya
     * @param userId Yükleyen kullanıcı ID
     * @return String dosya yolu
     */
    String addAttachment(Long ticketId, MultipartFile file, Long userId);

    /**
     * Müşteri memnuniyet puanı ekle
     * @param ticketId Ticket ID
     * @param rating Puan (1-5)
     * @param feedback Geri bildirim metni
     */
    void addSatisfactionRating(Long ticketId, int rating, String feedback);

    // ============ İSTATİSTİK ve ANALİTİK ============

    /**
     * Şirket bazında ticket istatistikleri
     * @param companyId Şirket ID
     * @return Map<String, Object> istatistik verileri
     */
    Map<String, Object> getTicketStats(Long companyId);

    /**
     * Agent performans metrikleri
     * @param companyId Şirket ID
     * @param startDate Başlangıç tarihi
     * @param endDate Bitiş tarihi
     * @return List<Object[]> performans verileri
     */
    List<Object[]> getAgentPerformanceMetrics(Long companyId, LocalDateTime startDate, LocalDateTime endDate);

    /**
     * SLA ihlali kontrolü (scheduled task için)
     */
    void checkSlaViolations();

    /**
     * Ticket oluşturma trendi
     * @param companyId Şirket ID
     * @param days Kaç günlük trend
     * @return List<Object[]> trend verileri
     */
    List<Object[]> getTicketCreationTrend(Long companyId, int days);

    // ============ APPROVAL WORKFLOW METHODS ============

    /**
     * Ticket'ı RESOLVED yapıp MANAGER onayına gönder
     * @param id Ticket ID
     * @param resolutionSummary Çözüm özeti
     * @param userId İşlemi yapan kullanıcı ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse resolveTicketForApproval(Long id, String resolutionSummary, Long userId);

    /**
     * MANAGER onayı - SUPER_ADMIN onayına gönder
     * @param id Ticket ID
     * @param approvalComment Onay yorumu
     * @param managerId Manager ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse approveByManager(Long id, String approvalComment, Long managerId);

    /**
     * SUPER_ADMIN final onayı - CLOSED yapar
     * @param id Ticket ID
     * @param finalComment Final yorum
     * @param adminId Admin ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse approveByAdmin(Long id, String finalComment, Long adminId);

    /**
     * Onayı reddet - önceki duruma geri döndür
     * @param id Ticket ID
     * @param rejectionReason Red sebebi
     * @param userId İşlemi yapan kullanıcı ID
     * @return TicketDetailResponse
     */
    TicketDetailResponse rejectApproval(Long id, String rejectionReason, Long userId);
}